### PR TITLE
refactor(v6): remove instance `type` property + `isType`

### PR DESCRIPTION
**Pull request by ShaMan123**
## Motivation

Previous discussions about `type` in v6.
At the beginning we changed `type` to a getter (and added a spamming setter to warn apps of the change) returning the class name for 2 reasons:
- the class regisrty would pick it up without the dev needing to configure anything when creating a subclass
- because we wanted to save time and patch the change instead of going throw the code base

Then minifying made us use the static `type` property because it mangled class names.
However we didn't remove the `type` getter/setter, though we could have.
Apparently apps use the instance property. I am against.
The migration path is to set the static `type` value for serialization.
The rest is deprecated.

closes https://github.com/fabricjs/fabric.js/issues/9324

## Description

I have made fabric use only the static `type` property.

## Changes

BREAKING:
- rm `isType` - if a dev wants it they can create a util for it.
- rm `getObjects(...type)` signature => `getObjects()`
- Serializable fabric classes do not accept `type` as an option, it will be dicarded in the constructor. Again, devs can solve this problem or we can safegaurd all fromObject instead of the constructor - that sounds like a viable option

### CODE REVIEW

File: src/util/typeAssertions.ts
Comment (ShaMan123) [inline src/util/typeAssertions.ts:1]: I have PRs open to remove this bad design.
Why is it so bad? I subclass one of these classes - I MUST keep the type value the same so to for serialization to work as expected I will have to override `toObject`. This is an unneeded gotcha.

**Review by ShaMan123 — COMMENTED** [review]
Once all tests pass on d57969989 it will mean that fabric is indeed safe and I will remove the following safeguard
```
  get type() {
    throw new Error('get type');
  }

  set type(value) {
    throw new Error('set type');
  }
```

File: src/gradient/Gradient.ts [hunk 1]
```diff
@@ -193,16 +193,16 @@ export class Gradient<
...
-      offsetX += object.width / 2;
-      offsetY += object.height / 2;
+      offsetX += width / 2;
+      offsetY += height / 2;
     }
     // todo what about polygon/polyline?
-    if (isPath(object) && this.gradientUnits !== 'percentage') {
-      offsetX -= object.pathOffset.x;
-      offsetY -= object.pathOffset.y;
+    if (pathOffset && this.gradientUnits !== 'percentage') {
```
Comment (ShaMan123) [inline src/gradient/Gradient.ts:203]: what about this comment?

**Review by ShaMan123 — COMMENTED** [review]
tests should pass now

File: src/gradient/Gradient.ts (hunk 1 above)
Comment (ShaMan123) [reply src/gradient/Gradient.ts:203]: ` // todo what about polygon/polyline?`

**Review by ShaMan123 — COMMENTED** [review]
All tests have passed with the strict safeguard throwing errors - aa243ed
Now removed it

**Comment by ShaMan123** [comment]
Actually now we can discuss and decide if to allow devs to use `type` on an instance because fabric doesn't use it anymore and I wrote an extensive test covering all initialization usages.

In v5 changing the instance `type` prop would affect `toObject` and maybe even `fromObject` because it was on the prototype.
No it will not.
That is why I think we should not allow passing down `type`.
In such a case it is needed the app should override and subclass anyways so better we are strict about it.

**Comment by ShaMan123** [comment]
anyways it is easy to deal with now and should be done in a follow if we decide to.
So this is ready to merge

**Comment by asturur** [comment]
i m sure this wasn't the outcome of the discussion.
we can't remove the getter, nor kill an app because is trying to set something that doesn't have a setter.
so the getter and the setter should stay.

**Comment by asturur** [comment]
Where is the original chat? i can't find it anymore

**Comment by ShaMan123** [comment]
> i m sure this wasn't the outcome of the discussion. we can't remove the getter, nor kill an app because is trying to set something that doesn't have a setter. so the getter and the setter should stay.

If fabric doesn't use type anymore what diffreence does it make. It becomes as any property.
What was the outcome then?
At least we can agree that fromObject should remove the type property before passing options to constructor?
And why kill an app? the current situation kills apss that use type.

**Review by ShaMan123 — COMMENTED** [review]
x

**Comment by asturur** [comment]
Do you have a link to the old thread? is hard to remember what was said about this being an issue.

**Comment by ShaMan123** [comment]
https://github.com/fabricjs/fabric.js/issues/9324

File: src/util/typeAssertions.ts [hunk 2]
```diff
@@ -41,29 +40,36 @@ export const isCollection = (
...
 };
 
 export const isInteractiveTextObject = (
   fabricObject?: FabricObject
 ): fabricObject is IText | Textbox => {
   // we could use instanceof but that would mean pulling in Text code for a simple check
   // @todo discuss what to do and how to do
-  return !!fabricObject && fabricObject.isType('IText', 'Textbox');
+  return (
+    !!fabricObject &&
```
Comment (zhe-he) [inline src/util/typeAssertions.ts:68]: If I create a class TextBox2 that inherits from TextBox, it won't match here.  isPath and isTextObject also have the same issue. Would it be better to write it like this?
```
fabricObject instanceof IText && !(fabricObject instanceof Text)
```

Comment (ShaMan123) [reply src/util/typeAssertions.ts:68]: instanceof is the best solution not doubt
but it causes import cycles so we can't use it
this is part of the reason why I want to remove all these utils as done in #8950

**Comment by Smrtnyk** [comment]
so basically when we are subclassing and have custom toObject, we should do it like this
```typescript
toObject() {
    return {
        ...super.toObject(),
        type: "custom-type"
    }
}
```
so that when reviving serialized object it gets picked up by this type `custom-type` we set
but we should not set `type` class property anymore?

**Comment by ShaMan123** [comment]
That is correct
However the best practice is
```
class CustomObject extends FabricObject {
  static type = 'custom-type'
}
classRegistry.setClass(CustomObject)
```

It will handle both `toObject` and `fromObject` out of the box

**Pull request closed without merging**